skymp5-client/src/front/browser.ts

```diff
@@ -7,76 +7,85 @@ import {
   settings,
   Ui,
 } from "skyrimPlatform";
+import { DXScanCodes } from '../lib/dx-scan-codes';
+import { EventEmitter } from '../lib/event-emitter';
+import { badMenus } from '../lib/ui-menu';
+
+const getMultiBindigString = (...codes: DXScanCodes[]) => codes.sort((a, b) => a - b).join('-');
 
 export const main = (): void => {
-  const F2 = 0x3c;
-  const F6 = 0x40;
-  const Escape = 0x01;
-
-  const badMenus = [
-    "BarterMenu",
-    "Book Menu",
-    "ContainerMenu",
-    "Crafting Menu",
-    "GiftMenu",
-    "InventoryMenu",
-    "Journal Menu",
-    "Lockpicking Menu",
-    "Loading Menu",
-    "MapMenu",
-    "RaceSex Menu",
-    "StatsMenu",
-    "TweenMenu",
-  ];
+  const emitter = new EventEmitter();
 
-  browser.setVisible(false);
-  let visible = false;
   let noBadMenuOpen = true;
   let lastBadMenuCheck = 0;
 
+  let visible = false;
+  const browserSetVisible = (state: boolean) => {
+    visible = state;
+    browser.setVisible(state);
+  };
+  browser.setVisible(false);
   once("update", () => {
     visible = true;
     browser.setVisible(true);
   });
 
-  {
-    let pressedWas = false;
-
-    on("update", () => {
-      const pressed = Input.isKeyPressed(F2);
-      if (pressedWas !== pressed) {
-        pressedWas = pressed;
-        if (pressed) {
-          visible = !visible;
-        }
-      }
-
-      if (Date.now() - lastBadMenuCheck > 200) {
-        lastBadMenuCheck = Date.now();
-        noBadMenuOpen =
-          badMenus.findIndex((menu) => Ui.isMenuOpen(menu)) === -1;
-      }
-
-      browser.setVisible(visible && noBadMenuOpen);
-    });
-  }
-
-  {
-    let focused = false;
-    let pressedWas = false;
-
-    on("update", () => {
-      const pressed =
-        Input.isKeyPressed(F6) || (focused && Input.isKeyPressed(Escape));
-      if (pressedWas !== pressed) {
-        pressedWas = pressed;
-        if (pressed) {
-          focused = !focused;
-          browser.setFocused(focused);
-        }
-      }
-    });
-  }
+  let focused = false;
+  const browserSetFocused = (state: boolean) => {
+    focused = state;
+    browser.setFocused(state);
+  };
+
+  const inputChangeEvent = 'event:input-change';
+  const keyState: { num: number } = { num: 0 };
+
+  on('update', () => {
+    const numKeys = Input.getNumKeysPressed();
+
+    if (keyState.num !== numKeys) {
+      keyState.num = numKeys;
+      const keyCodes = Array(numKeys)
+        .fill(null)
+        .map((_, i) => Input.getNthKeyPressed(i));
```

Comment:
It seems to be too expensive to check the whole keyboard. Let's check only bound keyboard keys. Calling game functions is expensive.